include/private/algebra_vector.h

```diff
@@ -166,8 +166,8 @@ OSQPFloat OSQPVectorf_norm_inf_diff(const OSQPVectorf* a,
 /* ||v||2 */
 OSQPFloat OSQPVectorf_norm_2(const OSQPVectorf* v);
 
-/* mean of vector elements */
-OSQPFloat OSQPVectorf_mean(const OSQPVectorf* a);
+/* mean of vector elements - assumes all elements are positive */
```

Comment:
I don't understand what the assumptions are behind the `pos_mean` method here.   Is it "mean assuming all entries" are positive?  Is it "mean of the absolute values"?    If the latter, why not just take the one norm over the length?

Reply:
The assumption is that all entries in `a` are positive, which will allow the mean to be implemented one of two ways and get equivalent results:
1) Using the one norm divided by the length
2) Using the normal mean formula

Method 1 is done in the MKL and CUDA backend by using the `asum` class of BLAS functions followed by the divide by the length, and method 2 is used in the builtin backend because an equivalent 1-norm computation would involve the absolute value of the elements of the vector (so requires another operation/temp variable on each element). This actually was a discrepancy between the CUDA and MKL/builtin algebras before, with CUDA implicitly doing option 1 and MKL/builtin doing option 2, but all three were advertising they were doing the mean. So the idea behind this is to codify that assumption into the API, which also allows MKL to now call the BLAS function instead of doing a loop through all elements of the vector.

This assumption is valid in the place we use the mean - the matrix scaling computation - since it is taking the mean over the result of the infinity norms of the columns and the infinity norm is already defined to be always positive.

If we wanted to implement the true mean computation, it would involve writing a custom kernel to compute the sum of the vector entries in CUDA, and means we won't take advantage of the pre-made BLAS functions and their potential speed-up in either CUDA or MKL. Doing a true "absolute mean" (which is the mean over the absolute values of the vector elements) would similarly seem to slow down the builtin algebra computation due to the added absolute value operation.

Reply:
Is the issue here down to BLAS not providing a direct way to sum a vector?  If so, then I think it is possible to do it using the BLAS `dot` function to compute a sum of elements in a vector `x` by telling it to compute <x , y> and then specifying that `y = 1.0` with an increment of `0`.

In that way you could directly compute a bare sum in MKL or wherever to compute the actual mean, and then just use that everywhere.   It does not seem like good design to have the actual behaviour of the function change depending on the algebra backend, even if it is only ever supplied with nonnegative inputs.

Reply:
> Is the issue here down to BLAS not providing a direct way to sum a vector?

Basically, yes - BLAS provides no way to directly sum a single vector. My concern with the dot function is that we are then adding all those multiplication operations (one per addition), which I guess might be only 1-2 cycles on most machines, but that could add up for larger problems, which is what the MKL and CUDA backends target. If we truly don't want to rely on the API specifying the vector is all positive, then adding an abs to the builtin algebra is probably the better option.

Reply:
Isn't this mean function only used in the equilibration step anyway?   That's already an extremely small part of the total compute time, so I don't think any extra ops here matter.

If you're going to add an abs to the building algebra anyway, then at that point we might as well drop this function entirely and manually compute in the 1-norm / length in the one place (I think) it is used.

Reply:
I have gone ahead and made the algebra libraries compute the 1-norm using the asum functions (if available) and then the usage of the average actually computes the mean at the location it is needed.